virtool/uploads/api.py

```diff
@@ -53,59 +58,137 @@ async def naive_writer(req, file_id):
             size += len(chunk)
             await handle.write(chunk)
 
-    return size
+    return size, virtool.utils.timestamp()
 
 
-@routes.post("/upload/{file_type}", permission="upload_file")
+@routes.post("/api/uploads", permission="upload_file")
 async def upload(req):
-    db = req.app["db"]
-
-    file_type = req.match_info["file_type"]
-
-    if file_type not in FILE_TYPES:
-        return not_found()
+    db = req.app["postgres"]
+    upload_type = req.query.get("type")
 
     errors = naive_validator(req)
 
     if errors:
         return invalid_query(errors)
 
-    filename = req.query["name"]
+    name = req.query["name"]
+
+    if upload_type not in UPLOAD_TYPES:
+        return bad_request("Unsupported upload type")
+
+    upload_ = await virtool.uploads.db.create(db, name, upload_type, user=req["client"].user_id)
+
+    upload_id = upload_["id"]
 
-    document = await virtool.files.db.create(
-        db,
-        filename,
-        file_type,
-        user_id=req["client"].user_id
-    )
+    async with AsyncSession(db) as session:
+        try:
+            result = (await session.execute(select(Upload).filter_by(id=upload_id))).scalar()
+            size, uploaded_at = await naive_writer(req, upload_id, name)
 
-    file_id = document["id"]
+            result.size = size
+            result.name_on_disk = f"{upload_id}-{name}"
+            result.uploaded_at = uploaded_at
 
-    try:
-        size = await naive_writer(req, file_id)
+            await session.commit()
 
-        await db.files.update_one({"_id": file_id}, {
-            "$set": {
+            upload_.update({
                 "size": size,
-                "ready": True
+                "name_on_disk": f"{upload_id}-{name}",
+                "uploaded_at": uploaded_at
+            })
```

Comment:
In this block, `name_on_disk` is derived in three different places. Why isn't it done in `db.create()`? You can create the record and get its `name_on_disk` in one transaction.

The duplication between updating the model at line 88 and updating the `dict` at line 94 can be simplified using `flush()` as you did in `db.create()`.